multicluster/controllers/multicluster/node_controller.go

```diff
@@ -0,0 +1,175 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	mcsv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	"antrea.io/antrea/multicluster/controllers/multicluster/common"
+)
+
+type (
+	// NodeReconciler is for member cluster only.
+	// It will create a Gateway object if a Node has an annotation `multicluster.antrea.io/gateway:true`
+	// and update corresponding Gateway if any subnets changes.
+	NodeReconciler struct {
+		client.Client
+		Scheme     *runtime.Scheme
+		namespace  string
+		precedence mcsv1alpha1.Precedence
+	}
+)
+
+// NewNodeReconciler creates a NodeReconciler to watch Node object changes and create a
+// corresponding Gateway if the Node has the annotation `multicluster.antrea.io/gateway:true`.
+func NewNodeReconciler(
+	client client.Client,
+	scheme *runtime.Scheme,
+	namespace string,
+	precedence mcsv1alpha1.Precedence) *NodeReconciler {
+	if string(precedence) == "" {
+		precedence = mcsv1alpha1.PrecedencePrivate
+	}
+	reconciler := &NodeReconciler{
+		Client:     client,
+		Scheme:     scheme,
+		namespace:  namespace,
+		precedence: precedence,
+	}
+	return reconciler
+}
+
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=gateways,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch;
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=gateways/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=gateways/finalizers,verbs=update
+
+func (r *NodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.V(2).InfoS("Reconciling Node", "node", req.Name)
+	gw := &mcsv1alpha1.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: r.namespace,
+		},
+	}
+	// When the Node is annotated with 'multicluster.antrea.io/gateway=true' as a Gateway:
+	//   - Delete the Gateway if the Node is deleted
+	//   - Update the Gateway if Node's InternalIP or GatewayIP is updated
+	//   - Create a new Gateway if there is no existing Gateway
+	node := &corev1.Node{}
+	if err := r.Client.Get(ctx, req.NamespacedName, node); err != nil {
+		if apierrors.IsNotFound(err) {
+			err := r.Client.Delete(ctx, gw, &client.DeleteOptions{})
+			return ctrl.Result{}, client.IgnoreNotFound(err)
+		}
+		klog.ErrorS(err, "Failed to get Node", "node", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	_, isGW := node.Annotations[common.GatewayAnnotation]
+	var err error
+	var gwIP, internalIP string
+	gwNamespacedName := types.NamespacedName{
+		Name:      node.Name,
+		Namespace: r.namespace,
+	}
+
+	// TODO: cache might be stale, need to revisit here and other reconcilers to
+	// check if we can improve this with 'Owns' or other methods.
+	if err := r.Client.Get(ctx, gwNamespacedName, gw); err != nil {
+		if apierrors.IsNotFound(err) && isGW {
+			if internalIP, gwIP, err = r.getGatawayNodeIP(node); err != nil {
```

Comment:
Nit: might be simpler to define a boolean like "gwExists" and based on that to decide or update out of the block, and so linke 108 - 113 can be shared.